### ISSUE TITLE
CPU memory keeps increasing during FSDP2 rollout weight sync and eventually triggers Ray   OOM

### ISSUE THREAD

**Issue by jack0rich** [verl-project/verl#6468]
### System Info

----------Python Info----------
Version      : 3.11.15
Compiler     : GCC 14.3.0
Build        : ('main', 'Mar 11 2026 17:20:07')
Arch         : ('64bit', 'ELF')
------------Pip Info-----------
Version      : 26.1.1
Directory    : /data/conda/qwen35-verl/lib/python3.11/site-packages/pip
vllm         : 0.19.1
sglang       : not found.
ray          : 2.55.1
torch        : 2.10.0
----------verl Info-----------
Version      : 0.8.0.dev
----------Platform Info----------
Platform     : Linux-5.15.0-139-generic-x86_64-with-glibc2.31
system       : Linux
node         : dell-PowerEdge-T640
release      : 5.15.0-139-generic
version      : #149~20.04.1-Ubuntu SMP Wed Apr 16 08:29:56 UTC 2025
----------Environment----------
CUDA Runtime : 12.8
CUDA Compiler : Cuda compilation tools, release 12.4, V12.4.131
----------System Info----------
CPU Memory      : 251.36 GB
GPU Count       : 4
GPU 1   Type    : NVIDIA GeForce RTX 3090
GPU 1   Memory  : 24.00 GB
GPU 2   Type    : NVIDIA GeForce RTX 3090
GPU 2   Memory  : 24.00 GB
GPU 3   Type    : NVIDIA GeForce RTX 3090
GPU 3   Memory  : 24.00 GB
GPU 4   Type    : NVIDIA GeForce RTX 3090
GPU 4   Memory  : 24.00 GB

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

 qwen3.5-2B DAPO

### Expected behavior

I observed a reproducible CPU memory growth issue during DAPO training with async rollout and FSDP2. The node memory keeps increasing across training steps and eventually Ray kills a worker due to OOM.

The OOM consistently happens after actor parameter update / rollout weight synchronization, before or around the next rollout. Ray object store usage is near zero, and simulator / environment workers do not show comparable linear growth.
At the OOM point, Ray reports the largest relevant processes as:
  ray::TaskRunner.run
  ray::WorkerDict.actor_rollout_ref_update_weights
  ray::WorkerDict.actor_rollout_ref_update_weights
  ray::WorkerDict.actor_rollout_ref_update_weights
  ray::VLNCETurnAgentLoopWorker
  VLLM::EngineCore

The killed worker may be a rollout/agent worker, but the active high-memory WorkerDict tasks are in actor_rollout_ref_update_weights.
The suspicious path appears to be FSDP2 rollout weight sync:
  WorkerDict.update_weights()
    -> actor.engine.get_per_tensor_param()
    -> FSDP2 DTensor full tensor materialization
    -> rollout.update_weights(...)

In particular, when VERL_FSDP2_WEIGHT_SYNC_GLOO=1 is enabled, _dtensor_full_tensor_gloo() gathers DTensor shards through CPU/Gloo:

  local_tensor = dtensor.to_local().detach().cpu().contiguous()
  torch.distributed.all_gather_object(...)
  full_tensor = torch.cat(...)

This seems to create CPU-side staging tensors/buffers whose memory is not fully released across steps, causing the WorkerDict RSS / node memory baseline to increase monotonically.

  Object store memory is negligible at OOM, so this does not look like a Ray object store leak.

  Expected behavior: CPU memory should return to a stable baseline after each rollout weight sync.

  Actual behavior: CPU memory keeps increasing across steps until Ray OOM killer terminates a worker.

**Comment by cben484** [comment]
I’m seeing the same issue. How much memory leaks per step in your case? In mine, it’s about 6.3 GiB per step.

**Comment by jack0rich** [comment]
We see the same leak scaling with model size: ~0.6 GiB/step for qwen3.5-2B and ~5.3 GiB/step for qwen2.5-3B. which model are you using?

**Comment by cben484** [comment]
In our experiments, we use Qwen3-35B.

**Comment by aoshen02** [comment]
still has such a problem when using the latest verl vllm image?
